### PR TITLE
discovery api for getting hostports by service name

### PR DESCRIPTION
**Pull request by ShanniLi**
r @jcorbin @kriskowal @anson627  
cc @blampe

### CODE REVIEW

File: node/hyperbahn/handler.js [hunk 1]
```diff
@@ -24,10 +24,14 @@
 
 var assert = require('assert');
 var util = require('util');
+var fs = require('fs');
+var path = require('path');
 var setTimeout = require('timers').setTimeout;
 
 var Errors = require('../errors.js');
 var TChannelJSON = require('../as/json');
+var TChannelThrift = require('../as/thrift');
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Let's not use thrift until `thriftify` is gone. I want to use the new `thriftrw` only.

File: node/hyperbahn/handler.js [hunk 2]
```diff
@@ -338,3 +359,108 @@ function unadvertise(service) {
     var self = this;
     self.channel.topChannel.handler.removeServicePeer(service.serviceName, service.hostPort);
 };
+
+HyperbahnHandler.prototype.handleHostportsByService =
+function handleHostportsByService(self, req, arg2, arg3, cb) {
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: req, head, body

File: node/hyperbahn/handler.js [hunk 3]
```diff
@@ -338,3 +359,108 @@ function unadvertise(service) {
...
+    var serviceName = arg3.serviceName;
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Cannot assume `serviceName` exists because lolthrift

File: node/hyperbahn/handler.js [hunk 4]
```diff
@@ -338,3 +359,108 @@ function unadvertise(service) {
...
+    var exitNodes = self.egressNodes.exitsFor(serviceName);
+    var exitHosts = Object.keys(exitNodes);
+
+    var myHost = self.channel.hostPort;
+    if (exitHosts.indexOf(myHost) === -1) {
+        // Since Hyperbahn is fully connected to service hosts,
+        // randomly picking one suffices
+        var exit = Math.floor(Math.random() * exitHosts.length);
+        self.sendRelayAsThrift({
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Just inline the request here. no need to retry

File: node/hyperbahn/handler.js [hunk 5]
```diff
@@ -338,3 +359,108 @@ function unadvertise(service) {
...
+        }, onFinish);
+    } else {
+        var svcchan = self.channel.topChannel.subChannels[serviceName];
+        var hosts = [];
+        if (svcchan) {
+            hosts = svcchan.peers.keys();
+        }
+        var err;
+        if (hosts.length === 0) {
+            err = NoHostError({
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Don't bother; just return an empty array.

File: node/hyperbahn/handler.js [hunk 6]
```diff
@@ -338,3 +359,108 @@ function unadvertise(service) {
...
+                serviceName: serviceName
+            });
+        }
+
+        onFinish(null, {
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Don't do this stuff. Use a different path for error and not error.

File: node/hyperbahn/handler.js [hunk 7]
```diff
@@ -338,3 +359,108 @@ function unadvertise(service) {
...
+    if (exitHosts.indexOf(myHost) === -1) {
+        // Since Hyperbahn is fully connected to service hosts,
+        // randomly picking one suffices
+        var exit = Math.floor(Math.random() * exitHosts.length);
+        self.sendRelayAsThrift({
+            hostPort: exitHosts[exit],
+            body: arg3,
+            inreq: req,
+            endpoint: 'Hyperbahn::hostportsByService'
+        }, onFinish);
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: dont pass the callback to the method. Do all the body building in here.

File: node/hyperbahn/handler.js [hunk 8]
```diff
@@ -338,3 +359,108 @@ function unadvertise(service) {
...
+                (
+                    codeName === 'NetworkError' ||
+                    codeName === 'Timeout'
+                )
+            ) {
+                setTimeout(tryRequest, self.relayAdRetryTime);
+            } else {
+                self.logError(err, opts, response);
+
+                callback(null, {
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Don't do boolean logic in the object. use proper if statements.

File: node/hyperbahn/hyperbahn.thrift [hunk 9]
```diff
@@ -0,0 +1,16 @@
+exception Exception {
+    1: required string message
+}
+
+exception NoHost {
+    1: required string message
+}
+
+service Hyperbahn {
+    list<string> hostportsByService(
```
Comment (Raynos) [inline node/hyperbahn/hyperbahn.thrift]: This is not how you do thrift. Always take one argument; a struct. Always return one argument a struct.

File: node/hyperbahn/hyperbahn.thrift [hunk 10]
```diff
@@ -0,0 +1,16 @@
+exception Exception {
+    1: required string message
+}
+
+exception NoHost {
```
Comment (Raynos) [inline node/hyperbahn/hyperbahn.thrift]: These exceptions are useless; they need more fields; namely `type` and `fullType` and any other meta info.

File: node/hyperbahn/hyperbahn.thrift [hunk 11]
```diff
@@ -0,0 +1,16 @@
+exception Exception {
```
Comment (Raynos) [inline node/hyperbahn/hyperbahn.thrift]: A catch all exception is useless.

File: node/hyperbahn/hyperbahn.thrift (hunk 9 above)
Comment (Raynos) [inline node/hyperbahn/hyperbahn.thrift]: The logical operation here is `discover`. Let's call this `discover`

**Comment by Raynos** [comment]
Let's shelf this feature until we are ready to fully commit to thrift.

File: node/hyperbahn/handler.js [hunk 12]
```diff
@@ -338,3 +359,108 @@ function unadvertise(service) {
...
+        if (hosts.length === 0) {
+            err = NoHostError({
+                serviceName: serviceName
+            });
+        }
+
+        onFinish(null, {
+            ok: !err,
+            body: err || hosts,
+            typeName: err ? 'noHost' : null
```
Comment (anson627) [inline node/hyperbahn/handler.js]: what is this typeName? error type?

File: node/hyperbahn/handler.js [hunk 13]
```diff
@@ -338,3 +359,108 @@ function unadvertise(service) {
...
+
+HyperbahnHandler.prototype.sendRelayAsThrift =
+function sendRelayAsThrift(opts, callback) {
+    var self = this;
+
+    var attempts = 0;
+
+    tryRequest();
+
+    // TODO: move functions out to methods
```
Comment (anson627) [inline node/hyperbahn/handler.js]: why not move it now?

File: node/hyperbahn/handler.js (hunk 1 above)
Comment (anson627) [inline node/hyperbahn/handler.js]: I guess I don't understand why thrift, even json dependency is already there.

Comment (ShanniLi) [inline node/hyperbahn/handler.js]: This is a task that is due next week ... And it specifically requires using thrift.

File: node/hyperbahn/handler.js (hunk 5 above)
Comment (ShanniLi) [inline node/hyperbahn/handler.js]: Empty array won't show up in TChannel stats. NoHostError will.

File: node/hyperbahn/hyperbahn.thrift (hunk 9 above)
Comment (ShanniLi) [inline node/hyperbahn/hyperbahn.thrift]: "This is not how you do thrift. Always take one argument; a struct. Always return one argument a struct."
this sounds wired. Any reason?

Comment (blampe) [inline node/hyperbahn/hyperbahn.thrift]: Extensibility in addition to weirdness around how Thrift handles function arguments. Similar in spirit to always returning JSON objects instead of lists.

File: node/hyperbahn/handler.js (hunk 1 above)
Comment (blampe) [inline node/hyperbahn/handler.js]: thriftify/thriftrw is just a matter of how the objects get serialized on the wire, no? Can we not swap this out with thriftrw later?

File: node/hyperbahn/handler.js (hunk 5 above)
Comment (Raynos) [inline node/hyperbahn/handler.js]: good point; seems legit.

File: node/hyperbahn/handler.js (hunk 1 above)
Comment (Raynos) [inline node/hyperbahn/handler.js]: thriftify has very low test coverage and can cause hyperbahn routers to crash in edge cases.

thriftrw is almost ready. I'd rather not add a production dependency to our hyperbahn deployment with low test coverage.

Comment (ShanniLi) [inline node/hyperbahn/handler.js]: so, should we stick to JSON for now?

Comment (Raynos) [inline node/hyperbahn/handler.js]: If we cant wait then just get a JSON version out. I'd rather we wait

Comment (ShanniLi) [inline node/hyperbahn/handler.js]: W.r.t. thriftify, it has been used by production for months, so I wouldn't worry too much about the test coverage.

Comment (Raynos) [inline node/hyperbahn/handler.js]: Apache thrift has been used in production for months. 

The discovery feature isn't urgent. No need to rush. Let's just do it right. Thrift is for life, you get one shot at the interface.

**Comment by ShanniLi** [comment]
Talked to @Raynos offline. Will shelf this until @kriskowal finishes the thriftrw

**Comment by Raynos** [comment]
It's been 3 weeks. Thriftrw is not as close as I thought. I'll unblock this. 

We can add thriftrw in the future. Cc @kriskowal @jcorbin

**Comment by jcorbin** [comment]
What is blocking thriftrw from being usable?

**Comment by ShanniLi** [comment]
Sounds good. I will move forward with thriftify. I will have this PR updated today or tomorrow.

File: node/hyperbahn/handler.js (hunk 4 above)
Comment (ShanniLi) [inline node/hyperbahn/handler.js]: two reasons: 1) generalizing this interface for moving other Hyperbahn endpoints to thrift. 2) retry is necessary, e.g., when k-value is being updated.

File: node/test/hyperbahn/hostports.js [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+        function onResponse(err, res) {
+            assert.ok(!err, 'should be no errors');
+            assert.ok(res, 'should be a result');
+            assert.ok(!res.ok, 'result should be not ok');
+            assert.equals(res.body.message, 'no peer available for matt', 'error message as expected');
+            assert.end();
+        }
+    });
+
+    HyperbahnCluster.test.only('get host port as expected', {
```
Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: dont leave onlys

File: node/hyperbahn/handler.js [hunk 15]
```diff
@@ -36,6 +40,22 @@ var RELAY_AD_TIMEOUT = 500;
 
 module.exports = HyperbahnHandler;
 
+// TODO: should be part of Hyperbahn error file
+var TypedError = require('error/typed');
+var NoPeersAvailable = TypedError({
+    type: 'no-peers-asvailable',
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: s/asvailable/available/

Comment (Raynos) [inline node/hyperbahn/handler.js]: hyperbahn.no-peers-available

File: node/hyperbahn/handler.js [hunk 16]
```diff
@@ -337,3 +365,139 @@ function unadvertise(service) {
...
+
+        res.push(obj);
+    }
+
+    return res;
+}
+
+HyperbahnHandler.prototype.discover =
+function discover(self, req, head, body, cb) {
+    var serviceName = body && body.query && body.query.serviceName;
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: No need to do any of this if checking; 

Write tests and rely on thriftrw.

File: node/hyperbahn/handler.js [hunk 17]
```diff
@@ -337,3 +365,139 @@ function unadvertise(service) {
...
+    if (typeof serviceName !== 'string' || serviceName.length === 0) {
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Should always be string; but do write tests. 

Do check for non empty though :)

File: node/hyperbahn/handler.js (hunk 4 above)
Comment (Raynos) [inline node/hyperbahn/handler.js]: Dont retry here; the edge client will retry.

Comment (Raynos) [inline node/hyperbahn/handler.js]: Do not prematurely generalize.

File: node/hyperbahn/handler.js [hunk 18]
```diff
@@ -337,3 +365,139 @@ function unadvertise(service) {
...
+                body: {
+                    peers: hosts
+                }
+            });
+        }
+    }
+};
+
+HyperbahnHandler.prototype.sendRelayAsThrift =
+function sendRelayAsThrift(opts, callback) {
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Lets drop the retries.

File: node/test/hyperbahn/hostports.js [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+                    serviceName: 'matt'
+                }
+            },
+            onResponse
+        );
+        function onResponse(err, res) {
+            assert.ok(!err, 'should be no errors');
+            assert.ok(res, 'should be a result');
+            assert.ok(!res.ok, 'result should be not ok');
+            assert.equals(res.body.message, 'no peer available for matt', 'error message as expected');
```
Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: assert res.body.serviceName

Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: How do we know which error it is ? Is there a `res.typeName` ?

File: node/test/hyperbahn/hostports.js [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+                },
+                check
+            );
+        }
+
+        function check(err, res) {
+            assert.ok(!err, 'should be no errors');
+            assert.ok(res, 'should be a result');
+            assert.ok(res.ok, 'result should be ok');
+            assert.equals(res.body[0], bob.channel.hostPort, 'should get the expected hostPort');
```
Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: This cant be right at all.

The response should have a peers field and stuff.

File: node/test/hyperbahn/hostports.js [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var DebugLogtron = require('debug-logtron');
```
Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: We need more tests.

Let's add a bunch of tests for thrift edge cases.

 - test sending a request with a malformed thrift IDL where we send an empty body. Assert BadRequest
 - test sending a request with a malformed thrift IDL where we send a body with a query without the serviceName field. Assert BadRequest
 - test sending a request with correct thrift IDL where we send an empty string service Name. Assert NotOk application error
 - test sending a request with a malformed thrift IDL which has zero exceptions defined; Send a request with empty string serviceName. Assert that the client doesn't blow up and gets back a not ok response.

**Comment by ShanniLi** [comment]
All tests written with failing tests commented out. See issues:
https://github.com/uber/thriftrw/issues/54
https://github.com/uber/thriftrw/issues/55

File: node/hyperbahn/handler.js [hunk 22]
```diff
@@ -337,3 +365,119 @@ function unadvertise(service) {
...
 };
+
+function convertHosts(hosts) {
+    var res = [];
+    for (var i = 0; i < hosts.length; i++) {
+        var strs = hosts[i].split(':');
+        var obj = {
+            port: parseInt(strs[1])
+        };
+        strs = strs[0].replace('localhost', '127.0.0.1').split('.');
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: There should not be any localhost in here...

File: node/hyperbahn/handler.js [hunk 23]
```diff
@@ -337,3 +365,119 @@ function unadvertise(service) {
...
+
+        res.push(obj);
+    }
+
+    return res;
+}
+
+HyperbahnHandler.prototype.discover =
+function discover(self, req, head, body, cb) {
+    var serviceName = body.query.serviceName;
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: This is only save to do once we fixed the thriftrw bugs.

File: node/hyperbahn/handler.js [hunk 24]
```diff
@@ -337,3 +365,119 @@ function unadvertise(service) {
...
+                cn: self.callerName
+            },
+            retryLimit: 0,
+            parent: opts.inreq
+        }), opts.endpoint, null, opts.body, onResponse);
+    }
+
+    function onResponse(err, response) {
+        if (response) {
+            return callback(null, {
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: just callback(null, response)

File: node/hyperbahn/handler.js [hunk 25]
```diff
@@ -337,3 +365,119 @@ function unadvertise(service) {
...
+    if (exitHosts.indexOf(myHost) === -1) {
+        // Since Hyperbahn is fully connected to service hosts,
+        // randomly picking one suffices
+        var exit = Math.floor(Math.random() * exitHosts.length);
+        self.sendRelayAsThrift({
+            hostPort: exitHosts[exit],
+            body: body,
+            inreq: req,
+            endpoint: 'Hyperbahn::discover'
+        }, cb);
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Don't do this.

Add a real handler here.

File: node/hyperbahn/handler.js [hunk 26]
```diff
@@ -337,3 +365,119 @@ function unadvertise(service) {
...
+        if (response) {
+            return callback(null, {
+                ok: response.ok,
+                body: response.body,
+                typeName: response.typeName
+            });
+        }
+
+        self.logError(err, opts, response);
+        callback(null, {
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: This is not valid at all. Just callback(err)

File: node/hyperbahn/handler.js [hunk 27]
```diff
@@ -337,3 +365,119 @@ function unadvertise(service) {
...
+    function onResponse(err, response) {
+        if (response) {
+            return callback(null, {
+                ok: response.ok,
+                body: response.body,
+                typeName: response.typeName
+            });
+        }
+
+        self.logError(err, opts, response);
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Let's not log here but instead of log in the other method.

File: node/hyperbahn/handler.js [hunk 28]
```diff
@@ -337,3 +365,119 @@ function unadvertise(service) {
...
+    function onIdentified(err) {
+        if (err) {
+            return onResponse(err);
+        }
+
+        self.tchannelThrift.send(self.channel.request({
+            host: opts.hostPort,
+            serviceName: 'hyperbahn',
+            trace: false,
+            timeout: self.relayAdTimeout,
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Wrong timeout; use a new timeout.

File: node/hyperbahn/handler.js (hunk 25 above)
Comment (Raynos) [inline node/hyperbahn/handler.js]: Instead of passing in `cb` you want to do something like

function handleForward(err, resp) {
    if (resp) {
        return cb(null, resp)
    }

    logAThingy();
    sendErrorFrame(errors.classify(err), err.message)
}
```

File: node/test/hyperbahn/hostports.js [hunk 29]
```diff
@@ -0,0 +1,267 @@
...
+            if (err) {
+                assert.end(err);
+            }
+            assert.ok(!res.ok, 'should be not ok');
+            // TODO: the res.body.message is missing, need to assert on that
+            assert.end();
+        }
+    });
+
+    // HyperbahnCluster.test('malformed thrift IDL: an empty body', {
```
Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: We will want to fix these bugs before landing.

File: node/hyperbahn/handler.js (hunk 25 above)
Comment (ShanniLi) [inline node/hyperbahn/handler.js]: Not sure I understand: sendErrorFrame(errors.classify(err), err.message)?
1) what is sendErrorFrame?
2) errors.classify is for tchannel error only. The error here can also be NoPeersAvailable or InvalidServiceName.

Comment (ShanniLi) [inline node/hyperbahn/handler.js]: kind of understand what you mean. I will send out an update.

File: node/test/hyperbahn/hostports.js (hunk 29 above)
Comment (ShanniLi) [inline node/test/hyperbahn/hostports.js]: These are minor issues of thriftrw and only occur under bad IDLs (very uncommon scenarios). They are also orthogonal to this PR. Can we land this PR first? :)

File: node/hyperbahn/handler.js (hunk 16 above)
Comment (Raynos) [inline node/hyperbahn/handler.js]: This statement should be save to have as `var serviceName = body.query.serviceName` again.

File: node/hyperbahn/handler.js [hunk 30]
```diff
@@ -337,3 +369,128 @@ function unadvertise(service) {
...
+    }
+
+    var exitNodes = self.egressNodes.exitsFor(serviceName);
+    var exitHosts = Object.keys(exitNodes);
+
+    var myHost = self.channel.hostPort;
+    if (exitHosts.indexOf(myHost) === -1) {
+        // Since Hyperbahn is fully connected to service hosts,
+        // randomly picking one suffices
+        var exit = Math.floor(Math.random() * exitHosts.length);
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Wait; why are we picking a random peer ? 

We can just make a request on the sub channel for the serviceName and sub channel will pick a random peer for us.

We can also skip `waitForIdentified` that way and use retries on other random peers :D

File: node/hyperbahn/handler.js [hunk 31]
```diff
@@ -337,3 +369,128 @@ function unadvertise(service) {
...
+                ok: true,
+                body: {
+                    peers: hosts
+                }
+            });
+        }
+    }
+};
+
+HyperbahnHandler.prototype.sendRelayAsThrift =
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: We should not need this helper if we make a retryable request directly on the sub channel.

File: node/test/hyperbahn/hostports.js [hunk 32]
```diff
@@ -0,0 +1,271 @@
...
+            {
+                query: {
+                    serviceName: ''
+                }
+            },
+            onResponse
+        );
+        function onResponse(err, res) {
+            assert.ok(err, 'should be error');
+            assert.equals(err.type, 'tchannel-thrift-handler.parse-error.body-failed',
```
Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: Can we add an assertion that this is a response parsing failure ?

File: node/test/hyperbahn/hostports.js [hunk 33]
```diff
@@ -0,0 +1,271 @@
...
+        var badThrift = new TChannelAsThrift({source: badSource});
+        badThrift.send(request,
+            'Hyperbahn::discover',
+            null,
+            {query: {}},
+            onResponse
+        );
+        function onResponse(err, res) {
+            assert.ok(err, 'should be error');
+            assert.equals(err.type, 'tchannel.bad-request', 'error should be bad request');
```
Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: Can we add an assertion on the `err.message` to show why it failed ?

File: node/test/hyperbahn/hostports.js [hunk 34]
```diff
@@ -0,0 +1,271 @@
...
+        var badThrift = new TChannelAsThrift({source: badSource});
+        badThrift.send(request,
+            'Hyperbahn::discover',
+            null,
+            {},
+            onResponse
+        );
+        function onResponse(err, res) {
+            assert.ok(err, 'should be error');
+            assert.equals(err.type, 'tchannel.bad-request', 'error should be bad request');
```
Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: Can we add an assertion on the `err.message` to show why it failed ?

File: node/test/hyperbahn/hostports.js [hunk 35]
```diff
@@ -0,0 +1,271 @@
...
+            );
+        }
+
+        function check(err, res) {
+            if (err) {
+                assert.end(err);
+            }
+            assert.ok(res, 'should be a result');
+            assert.ok(res.ok, 'result should be ok');
+            console.log(res.body.peers[0]);
```
Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: no console.log

File: node/test/hyperbahn/hostports.js [hunk 36]
```diff
@@ -0,0 +1,271 @@
...
+
+module.exports = runTests;
+
+if (require.main === module) {
+    runTests(require('../lib/hyperbahn-cluster.js'));
+}
+
+function covertHost(host) {
+    var res = '';
+    console.log(host.ip.ipv4);
```
Comment (Raynos) [inline node/test/hyperbahn/hostports.js]: no console.log

Comment (ShanniLi) [inline node/test/hyperbahn/hostports.js]: haha, my bad. :)

File: node/hyperbahn/handler.js (hunk 30 above)
Comment (ShanniLi) [inline node/hyperbahn/handler.js]: Good point!

File: node/test/hyperbahn/hostports.js (hunk 33 above)
Comment (ShanniLi) [inline node/test/hyperbahn/hostports.js]: Will add. However, such validation should really be in thriftrw ...

File: node/hyperbahn/handler.js [hunk 37]
```diff
@@ -337,3 +370,98 @@ function unadvertise(service) {
...
+            if (err) {
+                self.channel.logger.error('Failed to call discover API on exit node', {
+                    error: err,
+                    serviceName: serviceName
+                });
+                return cb(err, null);
+            }
+
+            // Need to reconstruct the error
+            if (resp.body.name === 'ThriftException') {
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: This feels liks a really bad check. Can we do `resp.typeName` ?

File: node/hyperbahn/handler.js [hunk 38]
```diff
@@ -337,3 +370,98 @@ function unadvertise(service) {
...
+    if (exitHosts.indexOf(myHost) === -1) {
+        // Since Hyperbahn is fully connected to service hosts,
+        // any exit node suffices.
+        svcchan = self.channel.topChannel.handler.getOrCreateServiceChannel(serviceName);
+        self.tchannelThrift.send(svcchan.request({
+            serviceName: 'hyperbahn',
+            headers: {
+                cn: 'hyperbahn'
+            },
+            parent: req
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: Let's set a `timeout` and `timeoutPerAttempt`; consider 5000 & 500 for defaults.

Let's also set `trace: false`

**Comment by Raynos** [comment]
Looking good. #soclose

File: node/as/thrift.js [hunk 39]
```diff
@@ -323,7 +322,9 @@ TChannelAsThrift.prototype._parse = function parse(opts) {
         if (bodyRes.value && opts.ok) {
             bodyRes.value = bodyRes.value.success;
         } else if (bodyRes.value && !opts.ok) {
-            bodyRes.value = onlyProperty(bodyRes.value);
+            var typeName = onlyKey(bodyRes.value);
+            bodyRes.value = bodyRes.value[typeName];
+            bodyRes.value.typeName = typeName;
```
Comment (Raynos) [inline node/as/thrift.js]: dont mutate `bodyRes.value`. Just set `bodyRes.typeName`

**Comment by Raynos** [comment]
Looks good. Let's get this in after tests pass.

**Comment by Raynos** [comment]
#sick
